### PR TITLE
Mid Trade Drawdown (New incentive mechanism)

### PR DESCRIPTION
**Pull request by jbonilla-tao**
### Description
[Provide a brief description of the changes introduced by this pull request.]

### Related Issues (JIRA)
[Reference any related issues or tasks that this pull request addresses or closes.]

### Checklist
- [ ] I have tested my changes on testnet.
- [ ] I have updated any necessary documentation.
- [ ] I have added unit tests for my changes (if applicable).
- [ ] If there are breaking changes for validators, I have (or will) notify the community in Discord of the release.

### Reviewer Instructions
[Provide any specific instructions or areas you would like the reviewer to focus on.]

### Definition of Done
- [ ] Code has been reviewed.
- [ ] All checks and tests pass.
- [ ] Documentation is up to date.
- [ ] Approved by at least one reviewer.

### Checklist (for the reviewer)
- [ ] Code follows project conventions.
- [ ] Code is well-documented.
- [ ] Changes are necessary and align with the project's goals.
- [ ] No breaking changes introduced.

### Optional: Deploy Notes
[Any instructions or notes related to deployment, if applicable.]

/cc @mention_reviewer

### CODE REVIEW

File: vali_objects/vali_dataclasses/perf_ledger.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        if accumulated_time < 0:
+            bt.logging.error(f"Negative accumulated time: {accumulated_time} for miner {self.miner_hotkey}."
+                             f" start_time_ms: {self.start_time_ms}, now_ms: {now_ms}")
+            accumulated_time = 0
+        cp.accumulated_ms += accumulated_time
+        cp.last_update_time_ms = now_ms
+        cp.n_updates += 1
+
+    def compute_return_between_ticks(self, current_portfolio_value: float, prev_portfolio_value: float):
+        return 1.0 + (current_portfolio_value - prev_portfolio_value) / prev_portfolio_value
```
Comment (trdougherty) [inline vali_objects/vali_dataclasses/perf_ledger.py:45]: return current_portfolio_value / prev_portfolio_value

File: vali_objects/vali_dataclasses/perf_ledger.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    def update_returns(self, current_cp: PerfCheckpoint, current_portfolio_value: float):
+        delta_return = self.compute_return_between_ticks(current_portfolio_value, current_cp.prev_portfolio_ret)
+        if current_portfolio_value >= current_cp.prev_portfolio_ret:
```
Comment (trdougherty) [inline vali_objects/vali_dataclasses/perf_ledger.py:49]: if delta_return >= 1

Comment (trdougherty) [reply vali_objects/vali_dataclasses/perf_ledger.py:49]: although I think the logic would have been the same

File: vali_objects/vali_dataclasses/perf_ledger.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        for cp in self.cps:
+            ans *= cp.loss_product
+        return ans
+
+    def get_total_ledger_duration_ms(self):
+        return sum(cp.accumulated_ms for cp in self.cps)
+
+def update_all_perf_ledgers(miner_hotkey_to_portfolio_value: dict[str, float]):
+    # Called every time MDD Checker runs. Currently means every 15-20 seconds.
+    miner_hotkey_to_perf_ledgers = ValiBkpUtils.safe_load_dict_from_disk('perf_ledgers.json', {})
```
Comment (trdougherty) [inline vali_objects/vali_dataclasses/perf_ledger.py:88]: should we set now_ms above the for loop to have a consistent evaluation time

File: vali_objects/vali_dataclasses/perf_ledger.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    for miner_hotkey, perf_ledger in miner_hotkey_to_perf_ledgers.items():
+        now_ms = TimeUtil.now_in_millis()
+        current_cp = perf_ledger.get_or_create_latest_cp(now_ms)
+        current_portfolio_value = miner_hotkey_to_portfolio_value.get(miner_hotkey, None)
```
Comment (trdougherty) [inline vali_objects/vali_dataclasses/perf_ledger.py:92]: to be sure I've got this right - the checkpoints are going to look at the entirety of the portfolio value at every iteration? How much overhead would it bring to track a ledger for each tradepair

File: vali_objects/vali_dataclasses/perf_ledger.py (hunk 1 above)
Comment (jbonilla-tao) [reply vali_objects/vali_dataclasses/perf_ledger.py:45]: nice simplification

File: vali_objects/vali_dataclasses/perf_ledger.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+from pydantic import BaseModel
+
+from time_util.time_util import TimeUtil
+from vali_objects.utils.vali_bkp_utils import ValiBkpUtils
+
+TARGET_CHECKPOINT_DURATION_MS = 36000000  # 10 hours
+TARGET_LEDGER_WINDOW_MS = 2592000000  # 30 days
+
+class PerfCheckpoint(BaseModel):
```
[the remark this answers is not in the document]
Comment (jbonilla-tao) [reply vali_objects/vali_dataclasses/perf_ledger.py:12]: yes

File: vali_objects/utils/position_manager.py [hunk 6]
```diff
@@ -273,7 +273,7 @@ def handle_eliminated_miner(self, hotkey: str,
                     open_position.close_out_position(TimeUtil.now_in_millis())
                     if source_for_elimination:
                         open_position.orders[-1].price_sources.append(source_for_elimination)
-                    self.save_miner_position_to_disk(open_position)
+                    self.save_miner_position_to_disk(open_position, delete_open_position_if_exists=False)
```
[the remark this answers is not in the document]
Comment (jbonilla-tao) [reply vali_objects/utils/position_manager.py:276]: i think trade pairs but i may have removed that in the newest commit

**Pull request merged**